### ISSUE TITLE
SteamDeck input duplicated when docked

### ISSUE THREAD

**Issue by 8bitskull** [defold/defold#7436]
I am having some issues with Steam Deck when external input is being used, for example when the Deck is in a dock or an external controller is connected. The core of the issue is that input is duplicated.

One user found a similar issue on a Linux desktop. They found that disabling SteamInput solved the issue in that context. The same user then tried to reproduce the issue on the Steam Deck, and provided me with this:

_I tried on the Steam Deck to see how it behaves._

_I don’t have a “dock”, but I have a basic usb-c hub with an hdmi output which should be similar enough. And indeed, I have the same behaviour when I use a controller instead of the embedded controls : the menu skips a step, and the screens turns black when I try to enter any menu. However, I didn’t see any option to disable SteamInput on the Deck… so I don’t think it’s possible to use the same workaround as on desktop._

_It doesn’t happen with the Deck embedded controls, even when using an external monitor. In that case, it works fine._

_For SteamInput, I think Steam sets environment variables so SDL will ignore controllers except the virtual one created for SteamInput. I don’t know which engine your game is built on, does it rely on SDL for controller input ? If so, maybe it ignores the environment variables Steam sets, and we get one input for the “real” controller, and one for the virtual one._

The key here seems to be the last paragraph, and something about SDL. I am not familiar with it - seems to be a low level library for handling multiple things, including input. [https://wiki.libsdl.org](https://wiki.libsdl.org/)

**Comment by 8bitskull** [comment]
Still getting reports about this issue. Seems to not be isolated to Steam Deck, but only Linux so far. Latest report was Nobara Linux (fedora based) with PowerA Enhanced Controller.

**Comment by britzl** [comment]
Ok, so the Steamdeck somehow duplicates input when docked and using an external controller. Do you know what kind of input events that are received?

Let's say that I press button on a connected controller. Will Defold then get two calls to `on_input()` on the same frame? And will the `action` table be exactly the same?

What about the gamepad connected event? Do you get one or two? And what are the controller names?

**Comment by 8bitskull** [comment]
It might be the case that this issue (https://github.com/defold/defold/issues/8059) needs to be resolved before we can move forward on this one. I note that gamepads are not detected in Linux (which includes Steam Deck) in Defold versions 1.4.8 onwards.

I built a gamepad testing utility for 1.4.7 and send it to someone who is experiencing the duplicate input issue. Here is the output:

![image](https://github.com/defold/defold/assets/24439991/5ccc739b-4f1b-4f36-8e3a-90f7a4ebffcb)

1) **Connection data** (middle left)
This is a list of "gamepad_connected" on_input() events. "Microsoft X-Box 360 pad 0" is how the Steam Deck is normally named. When not docked this is all that is reported for Steam Deck. I can't explain why there are 4 entries (I would expect 2 perhaps). The test was conducted with a Dualshock controller which may sometimes register a trackpad as a separate controller, but I am not sure because I don't own one.

2) **Input data** (bottom left)
This is a list of the 10 latest on_input() events with an action_id. We have a frame_id variable that is incremented by 1 in each update(). So an entry in this list with the same frame id means that we are getting more than one on_input() for a given frame. As you can see, we get two identical input events.

In conclusion:

- We get multiple gamepad connected events
- We get two calls to on_input()
- The action tables appear identical

**Comment by britzl** [comment]
> 2\. As you can see, we get two identical input events.

Well, *almost* identical. The `gamepad` fields are different, and they map to the `gamepad` fields of a gamepad connected event.

What you need to do is to build a table of connected gamepads, and then assign the connected gamepad ids to your player(s) and then **only** care about input events with a `gamepad` field matching the connected gamepad you assigned to each player. Something like this:

```lua

function init(self)
	self.connected_gamepads = {}
	self.player_gamepad = nil
end

local function handle_connected_gamepad(self, action)
	print("Connected gamepad", action.gamepad_name, "with id", action.gamepad)
	self.connected_gamepads[action.gamepad] = action
	if not self.player_gamepad then
		self.player_gamepad = action.gamepad
		print("Assigned gamepad to player")
	end
end

local function handle_disconnected_gamepad(self, action)
	print("Disconnected gamepad", action.gamepad_name, "with id", action.gamepad)
	self.connected_gamepads[action.gamepad] = nil
	if self.player_gamepad == action.gamepad then
		self.player_gamepad = nil
		print("Player gamepad disconnected")
	end
end

function on_input(self, action_id, action)
	if action_id == hash("gamepad_connected") then
		handle_connected_gamepad(self, action)
	elseif action_id == hash("gamepad_disconnected") then
		handle_disconnected_gamepad(self, action)
	elseif action.gamepad and action.gamepad == self.player_gamepad then
		-- handle player gamepad input here
	end
end
```

**Comment by 8bitskull** [comment]
Right, I see what you are saying. I've drafted up something similar to your examples above (thank you) that I think will work in principle. It's a bit more agnostic - I only ever make single player games so what I've done is make my input module destroy gamepad input with the same action_id in the same frame (beyond the first).

I don't think I can make an update of my live game until https://github.com/defold/defold/issues/8059 is fixed, and I don't think I can revert back to 1.4.7, so I need to wait until https://github.com/defold/defold/issues/8059 is fixed before I can fully confirm this as resolved. Up to you whether you want to close this issue before then or not!

**Comment by britzl** [comment]
> I only ever make single player games so what I've done is make my input module destroy gamepad input with the same action_id in the same frame (beyond the first).

Well, as long as it works for you then it's all good!



> I don't think I can make an update of my live game until #8059 is fixed, and I don't think I can revert back to 1.4.7, so I need to wait until #8059 is fixed before I can fully confirm this as resolved

I'll talk to the team about this one.

**Comment by 8bitskull** [comment]
Having debugged the issue much futher, it seems to only happen when Steam Input is enabled. While you get duplicate input, it is identified by the gamepad id as mentioned above, so it does mean you can handle it through logic on the Lua side. Don't believe there is anything wrong on the Defold side so I am closing the issue.
